monitoring/v1/types.go

```diff
@@ -0,0 +1,357 @@
+// namespace.  OpenShift requires admins to use the AlertingRule resource rather
+// than the upstream type in order to allow better OpenShift specific defaulting
+// and validation, while not modifying the upstream APIs directly.
+//
+// You can find upstream API documentation for PrometheusRule resources here:
+//
+// https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+// +genclient
+// +k8s:openapi-gen=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+type AlertingRule struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec describes the desired state of this AlertingRule object.
+	// +kubebuilder:validation:Required
+	Spec AlertingRuleSpec `json:"spec"`
+
+	// status describes the current state of this AlertOverrides object.
+	//
+	// +optional
+	Status AlertingRuleStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// AlertingRuleList is a list of AlertingRule objects.
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+// +k8s:openapi-gen=true
+type AlertingRuleList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// items is a list of AlertingRule objects.
+	// +kubebuilder:validation:Required
+	Items []AlertingRule `json:"items"`
+}
+
+// AlertingRuleSpec is the desired state of an AlertingRule resource.
+//
+// +k8s:openapi-gen=true
+type AlertingRuleSpec struct {
+	// groups is a list of grouped alerting rules.  Rule groups are the unit at
+	// which Prometheus parallelizes rule processing.  All rules in a single group
+	// share a configured evaluation interval.  All rules in the group will be
+	// processed together on this interval, sequentially, and all rules will be
+	// processed.
+	//
+	// It's common to group related alerting rules into a single AlertingRule
+	// resources, and within that resource, closely related alerts, or simply
+	// alerts with the same interval, into individual groups.  You are also free
+	// to create AlertingRule resources with only a single rule group, but be
+	// aware that this can have a performance impact on Prometheus if the group is
+	// extremely large or has very complex query expressions to evaluate.
+	// Spreading very complex rules across multiple groups to allow them to be
+	// processed in parallel is also a common use-case.
+	//
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MinItems:=1
+	// +kubebuilder:validation:Required
+	Groups []RuleGroup `json:"groups"`
```

Comment:
I don't remember why we said `+kubebuilder:validation:MinItems:=1` . An object without any group could be useful if you want to "remove" rules without deleting the object.

Reply:
Why would you want to keep an empty object around? That's not the expected way of interacting with kube, normally you'd just remove the object

Reply:
What does an empty group achieve in prometheus? Wondering why you wouldn't just remove it? (Which is the sentiment we would expect here)

Reply:
I think I'm in favor of allowing empty rule groups. Besides @simonpasquier's argument about it being a valid prometheus config, I can see future use cases using empty groups even if they arguably rather hypothetical at this point.
Say we'd like to add flexibility in terms of how often rules are evaluated. We could provide predefined rule groups (with their own eval interval) and let a user move rule between them.
@JoelSpeed I have removed the `MinItems` requirement. Let me know if that is objectionable.

Reply:
As I said before, I've got no strong feeling about this. Sticking with the usual Kubernetes practices is a good argument for having `MinItems > 0`.

> Do you ever see a future where there is a sibling to groups within the spec? Meaning that groups would be optional and not necessarily need an item? ie, a valid configuration that needs no groups?

I can't think of anything concrete on top of my head. If such situation arises, I guess that we could lift the requirement and allow empty groups?

Reply:
> I can't think of anything concrete on top of my head. If such situation arises, I guess that we could lift the requirement and allow empty groups?

Yes, we can relax this in the future provided we know that we can update all consumers to understand the change, luckily, we should be the owners of the consumers here. You couldn't make it tighter in the future as that could mean there are stored objects that no longer make the spec. For now I'd be in favour of including it, then relaxing in the future if we find a need to